doc/rados/configuration/bluestore-config-ref.rst

```diff
@@ -51,6 +51,151 @@ To specify a WAL device and/or DB device, ::
+    $ vgcreate ceph-block-2 /dev/sdc
+    $ vgcreate ceph-block-3 /dev/sdd
+
+Now create the logical volumes for ``block``::
+
+    $ lvcreate -l 100%FREE -n block-0 ceph-block-0
+    $ lvcreate -l 100%FREE -n block-1 ceph-block-1
+    $ lvcreate -l 100%FREE -n block-2 ceph-block-2
+    $ lvcreate -l 100%FREE -n block-3 ceph-block-3
+
+We are creating 4 OSDs for the four slow spinning devices, so assuming a 200GB
+SSD in ``/dev/sdx`` we will create 4 logical volumes, each of 50GB::
+
+    $ vgcreate ceph-db-0 /dev/sdx
+    $ lvcreate -L 50GB -n db-0 ceph-db-0
+    $ lvcreate -L 50GB -n db-1 ceph-db-0
+    $ lvcreate -L 50GB -n db-2 ceph-db-0
+    $ lvcreate -L 50GB -n db-3 ceph-db-0
+
+Finally, create the 4 OSDs with ``ceph-volume``::
+
+    $ ceph-volume lvm create --bluestore --data ceph-block-0/block-0 --block.db ceph-db-0/db-0
+    $ ceph-volume lvm create --bluestore --data ceph-block-1/block-1 --block.db ceph-db-0/db-1
+    $ ceph-volume lvm create --bluestore --data ceph-block-2/block-2 --block.db ceph-db-0/db-2
+    $ ceph-volume lvm create --bluestore --data ceph-block-3/block-3 --block.db ceph-db-0/db-3
+
+These operations should end up creating 4 OSDs, with ``block`` on the slower
+spinning drives and a 50GB logical volume for each coming from the solid state
+drive.
+
+Sizing
+======
+Sizing for the different devices needed in Bluestore can be more complicated
+than what it used to be for Filestore. To properly size a performant Bluestore
+OSD, special care needs to be placed on sizing ``block.db`` (the place where
+rocksdb lives) when using a :ref:`mixed spinning and solid drive setup
+<bluestore-mixed-device-config>`.
+
+There is no one-size-fits-all for sizing the ``block.db`` device, which will
+depend on the cluster workload. ``block.db`` should be large enough so that
+RocksDB data doesn't fill up to the point where it is forced to place data on
+the slow device (a performance penalty that is difficult to come back from).
+
+The formula to determine proper sizing of ``block.db`` needs the following
+items:
+
+* Average metadata size per object
+* Average object size
+* Size of the ``block`` (data) device
+
+Which are then used in the following way::
+
+    block device size / object size = Total Objects
+    (Total Objects * Average metadata size) * 1.10 = block.db size
+
+The ``1.10`` multiplier is adding ``10%`` more to the final size to prevent
+being overly correct and getting data to spill over, incurring in the
+performance penalty.
+
+Sizing strategies for different workloads
+-----------------------------------------
+There are three distinct workloads that may affect sizing: RBD, RGW, and
+CephFS. The workloads cause different outcomes because the object sizes and
+average metadata sizes are drastically different, so what will work fine for
+RGW, will probably not be fine for RBD.
+
+In the case of mixed use cases, the safest bet will be to go with the workload
+with the largest ``block.db``.
+
+
```

Comment:
Note that all of the figures below are very much approximations based on one person's experience and are unlikely to match your workload precisely.

Reply:
True, but it's better than 1 GB!!!  Also, I think it should mention that both Cephfs and RGW chop up large files/objects into 4 MB chunks, so if your average file/object size is >> 4 MB, then 4 MB is a good estimate for RADOS object size, which makes the arithmetic simpler.

For existing (Filestore) clusters, we can just ask Ceph what the number of objects and the amount of space used is, and calculate average object size from that.

Finally, for Cephfs, the average metadata per object might increase to be more like RBD for files accessed with random writes >= bluestore_min_alloc_size, because Bluestore will then have many more extents to deal with, right?  For example, database files.  But for sequential access cases, the metadata per object should be low, more like RGW, correct?

@markhpc ?

Reply:
Anything is better than 1 GB :)  I don't think I'm disagreeing.  I'm just generally uncomfortable with this entire section.. I don't think users should be thinking about any of this, even if they are expert users.  Instead, we should suggest that they use as much SSD as they have.  Maybe with a ceiling of something like 30%... or maybe not.  But I think the simple message of "more is better" is probably enough?

Reply:
In fact, even the RGW case is misleading, because for RGW index pools the block.db usage is unbounded (could be the entire data set!).

I think the only reason *not* to do "more is better" is that we don't currently put data on the SSD.  But we may very well do that in the future (case in point, ttps://github.com/ceph/ceph/pull/23208).  And even if we leaned on dm-cache to do it, that detail would be hidden by ceph-volume ("hi ceph-volume, here is my SSD and HDD, do the right thing").

Reply:
Part of the problem here is that "as big as possible" leaves a lot of room for interpretation and can cause doubt about what is correct vs. what may be problematic. I agree that these formulas are still hard to digest, but they do give us some pointers.

If what we want is to prevent users from figuring out some number (while causing confusion) at one end, and at the other end just being naive with ceph-volume and creating as-large-as-possible LVs, could we perhaps recommend a minimum size?

Given these examples, the largest `block.db` is the RGW example using close to 4% of the `block` device size. Wouldn't that be a reasonable default? Something like:

    It is recommended that the block.db size be 4% of the size of the block device. For a 1TB block, that would be a 40GB block.db LV. Anything less than 4% may incur in a performance penalty.

We would still prevent the performance penalty, we would have something pretty clear, and we wouldn't confuse users with formulas and average metadata/object sizes

Reply:
Well, it will never really cause a performance penalty, since the SSD is always strictly faster than teh HDD.  I think 1-2GB is probably the real minimum as that's what required to get the wal on the SSD (for fast metadata commits).  In practice, what really happens is that performance is good while the OSD is mostly empty (all the metadata fits on the SSD, yay!) and then things slow down once it fills up enough to not fit.  But it's still (a lot) better than no SSD at all.

Reply:
Taking the entire device and split it into equal parts for `block.db` can be tricky. Perhaps having guidance/formula just as suggested is a useful approach. I assume we have warnings in Ceph to tell: "hey osd.0 block.db is almost running out of space, please expand the block" or "hey osd.0 block.db is running out of space, we now write metadata on the block drive". Since we do everything with LVM it should be trivial to increase the size of the block, or is it something impossible and the initial size is set in stone? What I'm afraid about the "as big as possible" approach is that users might just divide the SSD into equal parts. This has pros and cons:

* Pros: we might never run out of space and thus fewer things to care on the operator side
* Cons: we can not scale our OSDs (disk on the machine) without adding a new SS

Also, the `block.db` fillup rate might differ depending on which application is writing to the OSD (rgw/rbd/cephfs)

We (hopefully) have a good set of tools to monitor the status of each block.db/wal at any given time so operators can react quickly.

My 2 cents.

Reply:
@leseb we can expand `block.db` with LVM, but once data has spilled over to the slower device it isn't trivial to move it back (this is my understanding from talking to @bengland2 ). So monitoring would only allow you to destroy it and reprovision the OSD (!)

My concern is also the scaling of OSDs, the `batch` subcommand could (by default) create `block.db` as big as possible but would be nice if we could also carve out the size from the config option in ceph.conf that ceph-disk used. 

My two issues are that:

1) There is no clear guidance to what is needed ("as big as possible" works for some users, doesn't give a better resource allocation idea for advanced users)

2) We can't really say "2GB" because this is all tied to the size of the `block` size. A 10TB disk would see the spill over problem faster (without filling up the OSD). 

I'm fine if we want to do without the formula, but would love to find some agreement on a recommendation on size (in addition to "as large as possible" which I find unclear without a path for expansion)

Reply:
My problem is that "need" and "problem" are the wrong words here.  You don't "need" an SSD to be any particular size; it's just a question of how much it will help.  Similarly, the spill-over effect isn't really a "problem"--it's just that the SSD didn't provide as much benefit as it could have.

The 2GB minimum is independent of the block device... it's based on how big the rocksdb wal files are (currently 512MB, and we need 2, plus some room or slop and for the bluefs metadata journal).  This is almost entirely independent of the main device's size.  Basically anything bigger than that is better, and how much better depends on how big.

"Big enough" is so wildly dependent on workload and data set I don't know that we can give useful guidance.  I think that, like the idea that you might want ot add another OSD to the host, this is really a hardware question: how many slots are in my box, and what is the ratio of HDD:SSD I'm willing to pay for?  If you have 2 empty slots with no HDDs, then we could leave some space on the SSD for those.  (This seems like a useful feature of the batch-prepare but not required for an MVP.)

It feels to me like the guidance how to pick that ratio (or go all-SSD) is a higher level recommendation.  "For RGW clusters, for best performance we recommend a dedicated pool of SSDs for bucket indexes.  In real-world deployments the capacity ratio between objects and bucket metadata appears to range between 1:100 and 1:200" (I'm making that up, not  sure what the real number is, but we can easily get a ballpark from current deployments.)

"For block workloads, best performance comes from all SSD pools.  For hybrid pools that mix HDDs and SSDs, an SSD:HDD ratio between 1:5 and 1:8 seems to work well."  That isn't quite right for NVMe, though, because it's more about capacity, and NVMes are getting big.  So maybe 1 NVMe for 12 OSDs is okay.  Which means the probably bigger consideration is around reliability and the blast radius of an NVMe failure.

Regardless, I'm still skeptical that we should (1) get into the weeds about block.db size, and if we do, (2) be so prescriptive when the workload descriptions are so imprecise and (3) bluestore behavior may change in the future.

Reply:
What everybody seems to forget is that SSDs have something called "lifetime". It is eitehr expressed in Writes/per day or something... But it boils down to that you can only write a cell so may times.
Now if you overdimension the SSD, the internal software in the SSD controller will balance the writes over all cells.  Partitioning has nothing to do with it, under water the SSD does writes where it pleases.

So in effect if you have a 120GB SSD, but only "use" 40Gb from the disk-side of things, you have extended the SSDs lifetime by a factor of 3.... So overdimentioning is not really a waste.... It just extends lifetime. We have servers where we run thru SSDs every 1,5 year. Going to the double size has make them last almost 3 years.

So then it becomes a question of servicability <> pricing <> reliability.